src/nodes.coffee

```diff
@@ -2112,7 +2140,39 @@ exports.Param = class Param extends Base
     @reference = node
 
   isComplex: ->
-    @name.isComplex() or @value instanceof Call
+    # Should this parameter be declared in the function body, rather than in
+    # the parameter list? We don’t want to pull out any parameter that
+    # has a value, or else we break default values; but we want things like
+    # generator functions and iterators to evaluate in the correct order,
+    # which they might not do unless we pull them out into the function body.
+    if @name.isComplex()
+      yes
+    else if @value? and @value.isComplex()
+      if @value instanceof Value
+        unless @value.base.isComplex()
+          no
+        else
+          # Objects and arrays are always complex, but unless they contain any
+          # complex children, they’re safe to leave in the parameter list.
+          if @value.base instanceof Arr or @value.base instanceof Obj
+            hasComplexChild = no
+            @value.base.traverseChildren no, (node) ->
+              if node.isComplex()
+                # This can be further refined. An empty object will evaluate as
+                # not having any complex children, but as long as it has at least
+                # one property it will be considered complex.
```

Comment:
Could you give examples on ...

- things that this code does _not_ consider complex
- things that this code _does_ consider complex, but could be improved no to in the future?